### ISSUE TITLE
Add support for the new URLSession async/await APIs

### ISSUE THREAD

**Issue by vdka** [kean/Pulse#113]
Simple integration using `URLSessionProxyDelegate` on iOS 16.0 is creating the request in logs, but they are stuck "Pending". I did a little bit of debugging and it seems `urlSession(_:task:didCompleteWithError:)` isn't being called by the system.

# Reproduction

```swift
import SwiftUI
import Pulse
import PulseUI

struct API {
    let session: URLSession = {
        let config = URLSessionConfiguration.default
        let delegate = URLSessionProxyDelegate()
        return URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
    }()

    func fetchTodos() async throws -> Data {
        let url = URL(string: "https://jsonplaceholder.typicode.com/todos/1")!
        let (data, _) = try await session.data(from: url)
        return data
    }
}

struct ContentView: View {
    let api = API()

    @State var isLoading = false
    @State var data: Data?

    @State var showPulse = false

    var body: some View {
        VStack {
            if let data {
                Text("\(data.count) bytes of data")
            } else if isLoading {
                ProgressView("Loading...")
            } else {
                Text("No data")
            }

            Button("Fetch Todos", action: fetchTodos).buttonStyle(.borderedProminent)

            Button("Show PulseUI", action: { showPulse = true }).buttonStyle(.bordered)
        }
        .sheet(isPresented: $showPulse, content: {
            PulseUI.MainView()
        })
        .padding()
    }

    func fetchTodos() {
        isLoading = true
        Task {
            self.data = try? await api.fetchTodos()
            isLoading = false
        }
    }
}
```

**Comment by kean** [comment]
I'm pretty sure it used to work in the previous versions, but yes, I just tested it, and the new async/await APIs on URLSession such as [`data(for:)`](https://developer.apple.com/documentation/foundation/urlsession/3767352-data) are not supported. The issue is that almost none of the delegate methods get called for tasks created using these methods. Crucially, `urlSession(_:task:didCompleteWithError:)` doesn't.

**Comment by vdka** [comment]
I thought that might be the case so I tried to work around it with the following but it runs into the same issue, which makes me think the issue is more related to the new concurrency system then URLSession's concurrency methods themselves.
```swift
func fetchTodos() async throws -> Data {
    let url = URL(string: "https://jsonplaceholder.typicode.com/todos/1")!

    // let (data, _) = try await session.data(from: url)
    // return data
    return try await withCheckedThrowingContinuation { continuation in
        let task = session.dataTask(with: URLRequest(url: url)) { data, _, error in
            if let error {
                continuation.resume(throwing: error)
                return
            }
            continuation.resume(returning: data ?? Data())
        }
        task.resume()
    }
}
```

**Comment by stremblayiOS** [comment]
+1

**Comment by kean** [comment]
@vdka , the completion-based methods also don't call the delegate. There is probably a way to swizzle some of the `URLSession` method to make it work automatically.

**Comment by stremblayiOS** [comment]
I believe we have to use `URLSessionTaskDelegate` in order to get the call back. Some people are discussing the issue here: https://www.reddit.com/r/swift/comments/ptpxf7/swift_55_async_await_issue_with_delegates_using/

**Comment by alexchernokun** [comment]
The same "pending" issue appears if you use Combine `dataTaskPublisher`. It seems like dataTaskPublisher doesn't invoke the delegate methods both for `URLSessionTaskDelegate` and `URLSessionDataDelegate`. Maybe someone also have this issue?

**Comment by vdka** [comment]
I'd love to make a PR to fix this, however I am not likely to get the time. If anybody is interested in looking into this, I know [ProxymanApp/Atlantis](https://github.com/ProxymanApp/atlantis) appears to be unaffected. It appears to Swizzle internal Method's and so AppReview would likely not approve, however, could be offered as an option similar to the existing [Experimental](https://kean-docs.github.io/pulse/documentation/pulse/experimental/urlsessionproxy) integration approach

**Comment by dnKaratzas** [comment]
You can use [Method Swizzling](https://nshipster.com/method-swizzling/) technique to swizzle certain functions of `NSURLSession` to capture HTTP/HTTPS traffic on the fly.

Check [Atlantis](https://github.com/ProxymanApp/atlantis)
https://github.com/ProxymanApp/atlantis/blob/main/Sources/NetworkInjector.swift
https://github.com/ProxymanApp/atlantis/blob/main/Sources/NetworkInjector%2BURLSession.swift

**Comment by ffittschen** [comment]
I'm assuming the `URLSession.DataTaskPublisher` internally uses the completionHandler based `dataTask` method. [Apple's Docs](https://developer.apple.com/documentation/foundation/urlsession/1407613-datatask) state that if a completionHandler is set, the delegate methods will not be called.

**Comment by tahirmt** [comment]
For the completion handler case, is it possible to tell the pulse logger after the completion block is called that the request is complete?

**Comment by jgodonQobuz** [comment]
Hi @kean 

I've exactly the same problem with "new" Swift concurrency URLSession as requests are always shown as "Pending..." when they are done like this:

```
let (data, _) = try await session.data(from: url)
```
<img src="https://user-images.githubusercontent.com/114389226/219318878-0b6dc214-3006-497f-ad11-542a4d4a5973.png" height="500">

As this is not possible/acceptable for our iOS team to modify all requests in another way than modern Swift concurrency `try await session.data(from:)`, do you think that any fixes/refacto could be done in Pulse about this in a futur version (v3 ?) of your great and so useful tool ?

Thanks for your answer and again: Big up for all your work 👏

Have a nice day

Jeremie

**Comment by MrAsterisco** [comment]
Just adding my two experience here: if you're using `URLSession.dataTaskPublisher`, you can replace it with this extension and you'll at least get network call logs _(no metrics, though)_.

It relies on creating a _fake_ `dataTask` and then subscribing to events of the `dataTaskPublisher`, as the publisher itself doesn't expose its `dataTask` instance. 

```swift 
extension URLSession {
  func trackedDataRequest(for urlRequest: URLRequest, with networkLogger: NetworkLogger) -> AnyPublisher<DataTaskPublisher.Output, DataTaskPublisher.Failure> {
    let dataTask = dataTask(with: urlRequest)
    
    return dataTaskPublisher(for: urlRequest)
      .handleEvents(
        receiveSubscription: { _ in
          networkLogger.logTaskCreated(dataTask)
        },
        receiveOutput: { output in
          networkLogger.logDataTask(dataTask, didReceive: output.data)
        },
        receiveCompletion: { completion in
          switch completion {
          case .finished:
            networkLogger.logTask(dataTask, didCompleteWithError: nil)
          case let .failure(error):
            networkLogger.logTask(dataTask, didCompleteWithError: error)
          }
        }
      )
      .eraseToAnyPublisher()
  }
}
```

**Comment by alexdremov** [comment]
For me, solution was to transform codebase to use `URLSessionProtocol ` protocol. It used by default in API generated by openapi-generator

```swift
public protocol URLSessionProtocol {
    func dataTask(with request: URLRequest, completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
}
```

Then, Pulse can be easily integrated:

```swift
class URLSessionPulse: URLSessionProtocol {
    let session: URLSession
    let networkLogger: NetworkLogger
    
    init(session: URLSession) {
        self.session = session
        self.networkLogger = NetworkLogger()
    }
    
    func dataTask(with request: URLRequest, completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
        var task: URLSessionDataTask?
        let onReceive: (Data?, URLResponse?, Error?) -> Void = { (data, response, error) in
            if let task {
                if let data {
                    self.networkLogger.logDataTask(task, didReceive: data)
                }
                self.networkLogger.logTask(task, didCompleteWithError: error)
            }
        }
        
        task = session.dataTask(with: request) {data, response, error in
            onReceive(data, response, error)
            completionHandler(data, response, error)
        }
        if let task {
            self.networkLogger.logTaskCreated(task)
        }
        return task!
    }
}
```

**Comment by salmin-skelia** [comment]
According to docs, if you're specifying completion handler (use async version that uses it under the hood or use dataTaskPublished that uses completion handler as well) it will ignore data delegate [note section of docs](https://developer.apple.com/documentation/foundation/urlsessiondatadelegate). I wonder how it worked before?

**Comment by salmin-skelia** [comment]
> For me, solution was to transform codebase to use `URLSessionProtocol` protocol. It used by default in API generated by openapi-generator, and `URLSession` conforms to it, so Pulse can be easily removed in production
> 
> ```swift
> public protocol URLSessionProtocol {
>     func dataTask(with request: URLRequest, completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
> }
> ```
> 
> Then, Pulse can be easily integrated:
> 
> ```swift
> class URLSessionPulse: URLSessionProtocol {
>     let session: URLSession
>     let networkLogger: NetworkLogger
>     
>     init(session: URLSession) {
>         self.session = session
>         self.networkLogger = NetworkLogger()
>     }
>     
>     func dataTask(with request: URLRequest, completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
>         var task: URLSessionDataTask?
>         let onReceive: (Data?, URLResponse?, Error?) -> Void = { (data, response, error) in
>             if let task {
>                 if let data {
>                     self.networkLogger.logDataTask(task, didReceive: data)
>                 }
>                 self.networkLogger.logTask(task, didCompleteWithError: error)
>             }
>         }
>         
>         task = session.dataTask(with: request) {data, response, error in
>             onReceive(data, response, error)
>             completionHandler(data, response, error)
>         }
>         if let task {
>             self.networkLogger.logTaskCreated(task)
>         }
>         return task!
>     }
> }
> ```

More generic solution that uses `URLSessionDataDelegate` and checks cancellation that worked for me.

`extension URLSession {
    /// Allows to track `URLSessionDataDelegate` using closure based call.
    /// By default if you use async interface or `completionHandler` based interface,
    /// URLSession won't notify `URLSessionDataDelegate`.
    public func dataTask(for request: URLRequest) async throws -> (Data, URLResponse) {
        var dataTask: URLSessionDataTask?

        let onSuccess: (Data, URLResponse) -> Void = { (data, response) in
            guard let dataTask, let dataDelegate = self.delegate as? URLSessionDataDelegate else {
                return
            }
            dataDelegate.urlSession?(self, dataTask: dataTask, didReceive: data)
            dataDelegate.urlSession?(self, task: dataTask, didCompleteWithError: nil)
        }
        let onError: (Error) -> Void = { error in
            guard let dataTask, let dataDelegate = self.delegate as? URLSessionDataDelegate else {
                return
            }
            dataDelegate.urlSession?(self, task: dataTask, didCompleteWithError: error)

        }
        let onCancel = {
            dataTask?.cancel()
        }

        return try await withTaskCancellationHandler(operation: {
            try await withCheckedThrowingContinuation { continuation in
                dataTask = self.dataTask(with: request) { data, response, error in
                    guard let data = data, let response = response else {
                        let error = error ?? URLError(.badServerResponse)
                        onError(error)
                        return continuation.resume(throwing: error)
                    }
                    onSuccess(data, response)
                    continuation.resume(returning: (data, response))
                }
                dataTask?.resume()
            }
        }, onCancel: {
            onCancel()
        })
    }
}`

**Comment by vitys96** [comment]
@salmin-skelia where delegate is used? can you show all code please?

**Comment by salmin-skelia** [comment]
> @salmin-skelia where delegate (URLSessionDataDelegate) is used? can you show all code please?

What do you mean by `where`? You just pass URLSessionDataDelegate as delegate for your session like `URLSession(configuration: sessionConfiguration, delegate: sessionDelegate, delegateQueue: nil)` and `sessionDelegate` using snipped I liked will receive appropriate action if `sessionDelegate` will conform `URLSessionDataDelegate`.

**Comment by kean** [comment]
Hey, folks. Thanks for engaging in this issue and suggesting the solutions! 

Pulse was originally designed back in 2020 for delegate-based `URLSession` instances, but I recognize that these are less prevalent now. I'm going to address it with the following changes in the upcoming Pulse 5.0 release.

## 1. Swizzling

Pulse will now offer swizzling inspired by https://github.com/ProxymanApp/atlantis suggested by @vdka. You'll be able to enable it with one line of code:

```swift
NetworkLogger.enableProxy()`.
```

It will be added to a separate `PulseProxy` library to eliminate the risk of app review rejection for folks who do not use it.  

## 2. URLSessionProtocol

`PulseProxy` cannot be a complete solution because, unlike Atlantis, the goal of Pulse is to provide not just logging but also debugging capabilities such as mocking, breakpoints, and rewriting without the need for an HTTP proxy.  To address it, I will follow @alexdremov's suggestion and add a `URLSessionProtocol` type to its Pulse-specific implementation that will power all its features.

```swift
let session: URLSessionProtocol
#if DEBUG
session = NetworkLogger.URLSession(configuration: .default)
#else
session = URLSession(configuration: .default)
#endif
```

`NetworkLogger.URLSession`  supports `URLSession` convenience closure-based APIs, async/await APIs, and I'm currently working on the Combine support. It uses a delegate-based approach similar to what @lukeredpath suggested.


This is the optimal solution because it doesn't rely on Objective-C runtime and opens more possibilities in the future to add more debugging features to `NetworkLogger.URLSession` that are impossible to add in any other way.

> so Pulse can be easily removed in production

It is an important point, and in both cases it will be relatively easy to disable Pulse in production or eliminate it completely using `#if canImport(Pulse)`.

### 3. URLSessionProxyDelegate

The existing `URLSessionProxyDelegate` is used internally by `NetworkLogger.URLSession` and will continue to be available as a way to enable logging for `Alamofire` or other framework that use delegate-based `URLSession` setup and where you can't access the underlying `URLSession` instances.

<hr/>

Swift made implementing these things quite a bit harder, and the integration will have too many choices, but that seems to be the only way. It will continue to be modular enough to accommodate almost any system. Once you set it up once, you generally won't have to think about it again.

The suggested changes are available in the following PR: https://github.com/kean/Pulse/pull/285. I'm open to suggestions, especially about the API design.

**Comment by kean** [comment]
The PR (https://github.com/kean/Pulse/pull/285) is largely complete except for a few methods. I would appreciate if someone could take it for a spin in their project and see how well it works.

**Comment by kean** [comment]
The changes shipped in [Pulse 5.0](https://github.com/kean/Pulse/releases/tag/5.0.0).
